Lavaland: Caves, Cliffs and more.

Venuska1117 opened the pull request:
## What Does This PR Do
This PR add new structures inspired by Space Ruins, so lavaland will have more interesting generation.
On top of that, it add new acient rock type called "Hardened Volcanic Rock" and Hard Lavaland Floor, with are exclusive for now in creating caves.

## Why It's Good For The Game
By creating a lot of caves, cliffs and other structures, it will help making lavaland more interesting to explore.

## Objectives
They can be added with time or changed, so people can see how many got added/changed in process.


- [x] Create 'o' shaped small cave.
- [x] Create 's' shaped small cave.
- [x] Create 'x' shaped small cave.
- [x] Mark if all small caves are done. 
- [x] -
- [x] Create 'X' shaped big cave.
- [x] Create 'Y' shaped big cave.
- [x] Create 'S' shaped big cave.
- [x] Mark if all big caves are done. 
- [x] -
- [x] Create 'Horizontal' Canyon.
- [x] Create 'Vertical' Canyon.
- [x] Create 'Z' Shaped Canyon.
- [x] Mark if all canyons are done.
- [x] -
- [x] Create 3 (or more) cave complex(es)
- [x] Mark if all cave complex(es) are done.
- [x] -
- [x] Create new Rock Type to distinguish cave structure.
- [x] Create new Floor Type to distinguish cave structure.
- [x] Make Cave Complexes contain "geodes" _(geodes have low chance to contain clown/mime ore)_

_If someone want me to add something more or discuss about it, ping me on discord or something._

## Testing
Compiled, spawned, checked in storms, for now work as intended.

## Changelog
:cl:
add: Lavaland get more structures types: 6 Caves, 3 canyons, 3 cave complex(es).
add: New rock type "Hard Volcanic Rock" with require diamond tools/psionic tools to be destroyed.
add: New Caves protect from ash storms, they can be distiguished by searching for new rock variant.
/:cl:

hal9000PR commented:
```
add: Added new "Hardened Volcanic Rock".
add: Added new "lava_land_surface_hard" volcanic floor.
``` these arent player facing changes

Venuska1117 commented:
> ```
> add: Added new "Hardened Volcanic Rock".
> add: Added new "lava_land_surface_hard" volcanic floor.
> ``` these arent player facing changes
> ```

thanks i will remove that

Venuska1117 commented:
I have no f- clue how to implement this templates to new generation system provided by #21838 i am bad at code and this over my capabilities, so i close this PR.
